### ISSUE TITLE
Unable to detect single-letter words with set_wordbreaks_utf32()

### ISSUE THREAD

**Issue by joshs6502** [adah1972/libunibreak#16]
When I use set_word_breaks_utf32() to navigate through words, I see that single letter words (i.e. 'a' in English, '北' in Chinese, ...) disappear because they always evaluate to WORDBREAK_BREAK and are consequently indistinguishable from surrounding whitespace.  The following code demonstrates the issue:

```
#include <stdio.h>
#include "wordbreak.h"

int main(int argc, const char* argv[]) {
    int i;
    uint32_t text[] = { 'T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', ' ', 't', 'e', 's', 't', '.', '\n' };
    char breaks[1024];
    size_t length = sizeof(text) / sizeof(text[0]);
    set_word_breaks_utf32(text, length, "", breaks);
    for(i = 0; i < length; i++) putchar(text[i]);
    for(i = 0; i < length; i++) putchar(breaks[i] + '0');
    putchar('\n');
    return 0;
}
```

The output of this code shows clearly that the letter 'a' is indistinguishable from the surrounding whitespace:

```
    This is a test.
    1110010000111000
```

What can I do to ensure that the boundaries of single letter words are distinguishable in set_word_breaks_utf32() output?

**Comment by joshs6502** [comment]
It appears from [this](http://stackoverflow.com/a/40557526/5472899) answer that I need to look at the class of each character between consecutive `WORDBREAK_BREAK` markers to determine which ones may constitute single character words.  I believe that `get_char_wb_class()` and `struct WordBreakProperties wb_prop_default[]` would answer that for me, but neither that function nor the struct are publicly exposed.

Is it truly necessary for me to evaluate these character classes to identify single-word boundaries, and if so would it be possible to add `get_char_wb_class()` or similar to the visible API?

**Comment by tasn** [comment]
How are you using the get class? This sounds like a possibly incomplete/non-future-proof solution.

The section you referred to in StackOverflow [(section 4)](http://www.unicode.org/reports/tr29/tr29-27.html#Word_Boundaries) mentions a few use-cases for the word boundary algorithm, each with its own special rules. For example, if you want to jump to the next word (ctrl + right), you'd want to jump over quotation-marks and parenthesis as if they were part of the word, while if you double click to select, you'd want to take them into account, so they are not selected with the word.

Therefore, a solution in the returned breaks array is not feasible.

I'm open to adding helper functions to help users of our API, but the question is what to add? In the [EFL](https://www.enlightenment.org) we have a simple function to check if a char is a whitespace char, and we use that. This could be easily added, but is that useful enough? Maybe we also want an "is alphanumeric"? I'm not sure.

**Comment by tasn** [comment]
The issue you described is exactly what I meant by:
> How are you using the get class? This sounds like a possibly incomplete/non-future-proof solution.

In particular, more classes you probably need to account for:

```
WBP_Katakana,
WBP_Hebrew_Letter,
WBP_MidNumLet,
WBP_MidLetter,
WBP_MidNum,
WBP_Numeric,
WBP_ExtendNumLet,
```
The problem with relying on these classes is that they are internal to the algorithm and can change (and do change) between different versions of Unicode. This means, we can't expose them as API if we want to be API/ABI stable.

This is why I suggested we expose a helper function that does something predictable, like letting you know if a character is a whitespace, however this feels out of scope for this library. With that being said, I understand this is a common pattern, so I'm willing to add API to solve it, but I want the API to be "good".

In summary, we need to come up with an API that helps us achieve what we want to achieve without exposing internals or being too broad, that is providing more than we need. I would rather keep the API concise avoiding confusion and cases we need to consider. I'm even thinking in the lines of providing a set of functions to get the actual word boundaries.

Any comments?

**Comment by joshs6502** [comment]
I see your point about the additional character classes.

As a novice user of libunibreak I'm not well qualified to propose changes,
but I agree with your reasoning about the API.  Your suggestion of
providing a set of functions to get word boundaries would be ideal for the
user, and would seem to fully fit within the scope of the library.

**Comment by tasn** [comment]
Please let me know if you end up coming with a good idea.

I need to think about it a bit more. I'd rather not introduce unnecessary APIs. Hopefully I'll have a solution in the next few hours.

**Comment by adah1972** [comment]
For comparison purposes, linebreak.h exposes our standard, simple API, but linebreakdef.h does expose a more "private" API, with enum LineBreakClass, struct LineBreakContext, etc. You may consider doing something similar. People using the *def.h should take some risks, but we do need to strive to keep versioning correct (including using the libtool versioning convention).

**Comment by tasn** [comment]
@adah1972, see my replies above. I find it very risky to expose these ever-changing classes. Just now between 8.0.0 and 9.0.0 more classes were added, which would make any code using these classes potentially break if the system's libunibreak is updated. So I would rather not force all of the users of our API update their code every time we change things.

I'm still trying to figure out the best course of action.

**Comment by tasn** [comment]
Just a quick update: I haven't forgotten about this issue, it's just that I feel uncomfortable with all the alternatives. The solutions don't feel "right". I wonder how other implementations deal with it.

**Comment by joshs6502** [comment]
This doesn't answer your question, but while working on something unrelated I stumbled upon the Windows approach for this.  The relevant Windows library is apparently called Uniscribe, and it provides the [ScriptBreak()](https://msdn.microsoft.com/en-us/library/windows/desktop/dd319118(v=vs.85).aspx) function.

ScriptBreak() fills in an array of `SCRIPT_LOGATTR` structs which simultaneously identify permissible line breaks, breakable whitespace, valid caret positions (i.e. not between combining marks), valid word breaks, and invalid or undisplayable characters.

_I will note that at least for word navigation purposes, Windows acts as if word breaks only occur at the beginning of a word, which I find displeasing and possibly at odds with the standard.  I note also that the `SCRIPT_LOGATTR` results seem to apply to characters rather than the space between them, which no longer seems right now that I have seen libunibreak's way of doing things._

Here is an example of what ScriptBreak() supplies:

<code>SCRIPT_LOGATTR &nbsp; H E L L O &nbsp;    ي   ِ  وِ  ا  س&nbsp;  ُ  ي   &thinsp;&thinsp; W O R L D</code>
<code>SoftBreak&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0</code>
<code>WhiteSpace &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0</code>
<code>CharStop &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1</code>
<code>WordStop &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0</code>
<code>Invalid &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code>

**Comment by tasn** [comment]
Hey,

Sorry, I don't know when/if I'll ever get to this. As you said, the uniscribe solution doesn't feel right. ICU implements iterators, which I guess works but comes with a lot of baggage. I have no idea, and looking at my coming future, due to real-life, I don't think I'll have plenty of time to deal with it.

Feel free to still suggest ideas, maybe we could get something in at some point.

Sorry.

**Comment by adah1972** [comment]
Closing this stalled issue, as there are no activities going on.
